### PR TITLE
Add prechecker

### PR DESCRIPTION
**Pull request by fearthecowboy**

### CODE REVIEW

File: modelerfour/test/test-check.ts [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+      const { session, expected, unexpected } = await createTestSession<Model>(cfg, folder, ['openapi-document.json', 'expected.yaml'], []);
+
+      // process OAI model
+      const prechecker = await new QualityPreChecker(session).init();
+
+      // go!
+      await prechecker.process();
+
+      console.error(``);
```
Comment (daviwil) [inline modelerfour/test/test-check.ts]: Is this needed anymore?

**Review by daviwil — APPROVED** [review]
This will be useful!

File: modelerfour/test/test-check.ts (hunk 1 above)
Comment (fearthecowboy) [reply modelerfour/test/test-check.ts]: Nope.  Still cleaning up.

**Review by daviwil — APPROVED** [review]
New check looks good

**Comment by daviwil** [comment]
Any idea why this change would cause the reordering effects we see in the regression test failure? Nothing seems like a breaking change, but it'd be good to have an explanation.

**Comment by daviwil** [comment]
Regression fix looks good.

**Comment by pakrym** [comment]
Do we know how many existing swaggers adding these checks would break? I know the network management one stopped working because of the property override error. Would we have to add an override the same way we did for name conflicts?

Other than that the fix seems to works.

**Comment by fearthecowboy** [comment]
I'm running a check on the specs repo right now.

**Comment by fearthecowboy** [comment]
I get 281 errors across 47 projects.

The vast majority are duplicate name and property redelcaration 

I'll add the switch to generate `_AUTOGENERATED` names to this PR, which will reduce the errors quite significantly

**Comment by fearthecowboy** [comment]
-- I've updated the prechecker with some better checks:

-- moved check that updates the missing type:object  from m4 into the prechecker
-- the check for duplicate (virtually identitcal) schemas is in there, and will give a message and deduplicate automatically when it can. It can't handle circular references right now.
-- checks for properties being redeclared from the parent, and issues a warning and drops the redeclaration in the child.
-- checks for duplicate parent allOf references and gives an error (I will work on making that auto-correct later)

-- it also checks for a new phenomenon where people are using an inline allOf by itself instead of using a schema $ref  (this apparently is some really bad $(%@ing advice that people are getting to "work around" the prohibition of having description,etc beside a $ref  .. ( apparently swashbuckle generates this monstrosity too) -- the check I have turns these cases into a $ref when it can. (because we permit a few things to live beside the $ref)

**Comment by pakrym** [comment]
What's the status/plan for this PR? Are we adding flags as part of it?

**Comment by daviwil** [comment]
I'm considering adding the schema name suffixing behavior to a separate PR since this one is focused on prechecker rules. Hope to start on that tomorrow!

**Comment by daviwil** [comment]
Updated plan is that I'll add the duplicate name suffixing behavior to this PR tomorrow (or latest Monday) because there is some new logic here that will be needed to avoid the errors from being raised when processing specs with duplicate names.

@fearthecowboy and I have discussed the implementation details and I'll get started on it as soon as I can.

**Comment by joheredi** [comment]
Just a note to keep in mind for pre-checker, I was talking to @fearthecowboy  offline about the scenario when a swagger defines an empty `{type: object}` in allOf, M4 gives an AnySchema as one of the parents.

image

I'll be ignoring parents with AnySchema for now in Autorest.Typescript

[Here is a swagger ](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2019-05-05-preview/AlertsManagement.json#L2273 )that does this

**Comment by pakrym** [comment]
Related https://github.com/Azure/autorest.modelerfour/issues/255

**Comment by daviwil** [comment]
@lmazuel Can you try running the above build with parameters `--ignore-errors --modelerfour.deduplicate-schema-names` to see whether the issue is resolved for more of the specs?

**Comment by daviwil** [comment]
I've updated this PR with a couple changes that deviate from prior behavior:

- `--modelerfour.deduplicate-schema-names` has now been renamed to `--modelerfour.lenient-model-deduplication`
- `--modelerfour.lenient-model-deduplication` now turns the following error message from ERROR level to WARNING level:

   ```
   ERROR (PreCheck/DuplicateSchema): Duplicate Schema named ResourceScenario2 -- properties.id.readOnly: true => <none>,properties.id.$ref: "#/components/schemas/schemas:12" => <none>,properties.id.type: undefined => "integer",properties.id.format: undefined => "int32",properties.name.readOnly: true => <none>,properties.name.$ref: undefined => "#/components/schemas/schemas:34"
   ```

   @lmazuel and I decided to make this fix so that `--ignore-errors` wouldn't need to be passed just to relax the duplicate model detection.  Since this is a temporary measure, we don't want to have teams start to get used to running with `--ignore-errors` turned on all the time because they might miss other legitimate errors in the prechecker phase.

@pakrym can you also give this a try sometime on Thursday to see if it removes the need for your workaround?

**Comment by daviwil** [comment]
@fearthecowboy It seems that we now have 3 places where modelerfour will raise an error about duplicate object schemas:

- Prechecker - https://github.com/fearthecowboy/autorest.modelerfour/blob/add-prechecker/modelerfour/quality-precheck/prechecker.ts#L236
- Prenamer - https://github.com/fearthecowboy/autorest.modelerfour/blob/add-prechecker/modelerfour/prenamer/prenamer.ts#L390
- Checker - https://github.com/fearthecowboy/autorest.modelerfour/blob/add-prechecker/modelerfour/checker/checker.ts#L42

Would it make sense to only raise a *single* error from checker with a more complete message that explains that they need to update their duplicate schema to use `x-ms-client-name`?  Something like:

> There are two or more instances of an object schema with the name "DuplicatedSchema".  Please either add "x-ms-client-name" to the duplicate schemas to differentiate them or use a reference to share the same schema.

Also, it seems like the [`fixSchemaCollisions`](https://github.com/fearthecowboy/autorest.modelerfour/blob/add-prechecker/modelerfour/prenamer/prenamer.ts#L377) behavior exposed by the `resolve-schema-name-collisions` parameter might be duplicate behavior.  Should we remove it? At the very least should we remove the error that it throws so that the user doesn't see 2-3 errors for a duplicate object schema?

**Comment by fearthecowboy** [comment]
Ideally, the prenamer should get out of the error-message business. (neither checker was there when that started).

We're gonna need checks on both ends, since some kinds of checks are only possible after, and some we can catch before modeling.

Too many errors when using `--ignore-errors`  isn't something I'd worry too much about.

**Comment by fearthecowboy** [comment]
LGTM

**Comment by daviwil** [comment]
I've looked over the changes reported by the regression test run and it's merely ordering changes to fields in models, nothing that would have an actual effect on the operation of the generated code.  Everything else should be ready to go in this PR, merging it now!

**Pull request merged**